### PR TITLE
refactor cache usage of kubernetesenv adapter

### LINKED ISSUE #2046: refactor mixer/adapter/kubernetes/cache code
It seems cache used in `mixer/adapter/kubernetes/` (and `kubernetes2`) can be simplified. This issue is a placeholder for me to address this. This should be started after https://github.com/istio/istio/pull/1982 is merged.

### PR DESCRIPTION
**Pull request by jmuk**
This does two different things to simplify the code in kubernetesenv
adapter:
- cache can be significantly simplified by using Indexer.
  The cache controller would be simply a SharedIndexInformer with
  additional utility method of accessing the cached data.
- use of kubernetes fake library for testing. This way, the test
  does not have to fake the store. Instead it fakes the actual
  k8s API to see the test set of pods.

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #2046 

**Special notes for your reviewer**:

**Release note**:

```release-note
none
```

### CODE REVIEW

**Comment by jmuk** [comment]
rebased, now reviewable. @douglas-reid, @guptasu PTAL.

File: mixer/adapter/kubernetesenv/cache.go [hunk 1]
```diff
@@ -39,228 +35,70 @@ type (
...
-
-const debugVerbosityLevel = 4
+func ipIndex(obj interface{}) ([]string, error) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		return nil, errors.New("object is not a pod")
+	}
+	ip := pod.Status.PodIP
+	if ip == "" {
+		return nil, nil
```
Comment (douglas-reid) [inline mixer/adapter/kubernetesenv/cache.go]: should this be `return []string{}, nil` or is nil expected?

Comment (jmuk) [reply mixer/adapter/kubernetesenv/cache.go]: `nil` is actually okay and safe, as it behaves like an empty slice.

**Pull request merged**